File: 2022/julia/02.jl
# Advent of Code 2022, day 2
# (c) aichingert

module Day2

function part1(inp)
	return sum(map(n-> n[2] + (3 - (2 + n[1] - n[2]) % 3) % 3 * 3, inp))
end

function part2()
	return sum(map(n-> (n[1]+n[2]) % 3 + 1 + (n[2] - 1) % 3 * 3, inp))
end

function map_value(s::SubString) 
	if s == "A" || s == "X" return 1 
	elseif s == "B" || s == "Y" return 2 
	else return 3 end
end

inp = []

for l in eachline("../input/02")
	(lhs,rhs) = split(l, " ")
	push!(inp, (map_value(lhs), map_value(rhs)))
end

println("Part 1:" * string(part1(inp)))
println("Part 2:" * string(part2()))

end

File: 2022/rust/02.rs
// Advent of Code 2022, day 2
// (c) aichingert

fn part1(n: &Vec<(i32,i32)>) -> i32 {
    n.iter().map(|(lhs,rhs)| rhs + (3 - (2 + lhs - rhs) % 3) % 3 * 3).sum::<i32>()
}

fn part2(n: &Vec<(i32,i32)>) -> i32 {
    n.iter().map(|(lhs,rhs)| (lhs+rhs) % 3 + 1 + (rhs - 1) * 3).sum::<i32>()
}

fn parse() -> Vec<(i32,i32)> {
    std::fs::read_to_string("../input/02").unwrap().trim().lines().map(|l| {
        let (lhs,rhs) = l.split_once(' ').unwrap();
        (map_value(lhs),map_value(rhs))
    }).collect::<Vec<(i32,i32)>>()
}

fn map_value(val: &str) -> i32 {
    match val {
        "A" | "X" => 1,
        "B" | "Y" => 2,
        "C" | "Z" => 3,
        _ => panic!("Invalid input!"),
    }
}

fn main() {
    let n = parse();

    println!("Part 1: {}", part1(&n));
    println!("Part 2: {}", part2(&n));
}
